sdks/python/apache_beam/testing/notebooks/runner.py

```diff
@@ -0,0 +1,161 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Executes notebook tests in current enviornment.
+
+Example usage:
+  Run all specs in ./base/:   python runner.py base
+  Run single test spec:       python runner.py --single-test base/testing.spec
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import glob
+import itertools
+import os
+import sys
+from collections import OrderedDict
+
+import yaml
+
+import apache_beam as beam
+import nbformat
+from nbconvert.preprocessors import ExecutePreprocessor
+
+# import testlib
+
+IS_BEAM_DEV = False
+if "dev" in beam.__version__:
+  IS_BEAM_DEV = True
+
+
+def _skip_test(spec):
+  if IS_BEAM_DEV and "sql" in str(spec).lower():
+    return True
+  return False
+
+
+class Executor:
+  """Loads a single test spec, executes and compares target notebook."""
+  def __init__(self, spec_path, timeout_secs):
+    with open(spec_path, "r") as spec_f:
+      self.spec = yaml.safe_load(spec_f)
+    # notebook in spec relative to spec's location
+    self.dir = os.path.dirname(spec_path)
+    self.notebook_path = os.path.join(self.dir, self.spec["notebook"])
+    self.tests = self.spec["tests"]
+    self.timeout_secs = timeout_secs
+
+  def execute_tests(self):
+    """Executes notebook and compares to test spec.
+
+    Returns:
+      # of tests, # of errors, error_dict
+      where error_dict maps (cell number, test class, expected output) to string
+    """
+    # load and execute notebook using nbconvert
+    with open(self.notebook_path, "r") as nb_f:
+      nb = nbformat.read(nb_f, as_version=4)
+    ExecutePreprocessor.timeout = self.timeout_secs
+    ep = ExecutePreprocessor(allow_errors=True)
+    # Executes full notebook, result is classic JSON representation
+    exec_nb, _ = ep.preprocess(nb, {"metadata": {"path": self.dir + "/"}})
+    test_count = 0
+    error_count = 0
+    errors = OrderedDict()
+    code_cells = {}
+    for cell in exec_nb["cells"]:
+      if cell["cell_type"] == "code":
+        code_cells[cell["execution_count"]] = cell
+    # generate test cases for each line, substituting in the output
+    for cell_num in sorted(self.tests.keys()):
+      if cell_num not in code_cells:
+        test_count += 1
+        error_count += 1
+        errors[(cell_num, "", "")] = "Given cell does not exist."
+      else:
+        cell = code_cells[cell_num]
+        for test in self.tests[cell_num]:
+          cls, setup = list(test.items())[0]
+          test_count += 1
+          try:
+            getattr(sys.modules["testlib"], cls)(setup).check(cell)
```

Comment:
I don't love this pattern because it makes it harder to see/understand the relationship between these 2 modules (and you lose things like good typing and some nicer errors). e.g. if we decided to move testlib.py in the future, would we notice that this breaks things like this? How many modules do you think there will be? Could we just introduce a function like:

```
def get_test_case(cls: str, setup: Any) -> NotebookTestCase:
  if cls == "RegexMatch":
    return RegexMatch(setup)
  ...
  raise Exception(f"no module named {cls} found in testlib.py")
```